crates/corro-agent/src/agent/handlers.rs

```diff
@@ -401,6 +404,131 @@ fn processing_cost(change: &Changeset) -> usize {
+                Some(change) => {
+                    if let Changeset::EmptySet { versions, ts } = change.changeset {
+                        buf.entry(change.actor_id).or_insert(vec![]).push((versions.clone(), ts));
+                        cost += versions.len();
+                    } else {
+                        warn!("received non-emptyset changes in emptyset channel from {}", change.actor_id);
+                    }
+
+                    if cost >= max_changes_chunk && !buf.is_empty() {
+                        process = true;
+                    }
+                },
+                None => break,
+            },
+            _ = max_wait.tick() => {
+                process = true
+            },
+            _ = &mut tripwire => {
+                break;
+            }
+        }
+
+        if process {
+            let mut retain_keys = Vec::new();
+            for (actor, changes) in &buf {
+                match process_emptyset(agent.clone(), bookie.clone(), *actor, changes.clone()).await
+                {
+                    Ok(()) => {
+                        let total = changes.iter().fold(0, |acc, (v, _)| acc + v.len());
+                        cost -= total;
+                    }
+                    Err(e) => {
+                        warn!("encountered error when processing emptyset - {e}");
+                        retain_keys.push(*actor);
+                    }
+                }
+            }
+
+            buf.retain(|key, _| retain_keys.contains(key));
+        }
+    }
+}
+
+pub async fn process_emptyset(
+    agent: Agent,
+    bookie: Bookie,
+    actor_id: ActorId,
+    changes: Vec<(Vec<RangeInclusive<corro_types::base::Version>>, Timestamp)>,
+) -> Result<(), ChangeError> {
+    let mut conn = agent.pool().write_low().await?;
+
+    let booked = {
+        bookie
+            .write(format!(
+                "process_emptyset(booked writer, updates timestamp):{actor_id}",
+            ))
+            .await
+            .ensure(actor_id)
+    };
+    let mut booked_write = booked
+        .write(format!(
+            "process_emptyset(booked writer, updates timestamp):{actor_id}"
+        ))
+        .await;
+    let mut snap = booked_write.snapshot();
+
+    let tx = conn
+        .immediate_transaction()
+        .map_err(|source| ChangeError::Rusqlite {
+            source,
+            actor_id: None,
+            version: None,
+        })?;
+
+    counter!("corro.sync.empties.count", "actor" => format!("{actor_id}"))
+        .increment(changes.len() as u64);
+    let mut max_ts: Timestamp = Default::default();
+    for (versions, ts) in changes {
+        debug!(self_actor_id = %agent.actor_id(), "processing emptyset changes, len: {}", versions.len());
+        for version in versions {
+            store_empty_changeset(&tx, actor_id, version.clone(), ts)?;
+        }
+        if ts > max_ts {
+            max_ts = ts;
+        }
+    }
```

Comment:
should we use one transaction per emptyset? this would make the transactions much smaller and easier to digest for the database.

Reply:
that makes sense.